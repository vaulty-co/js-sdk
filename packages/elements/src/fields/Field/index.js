import invariant from 'invariant';
import { SlaveChannel } from '@js-sdk/utils/src/channels/SlaveChannel';
import { Message } from '@js-sdk/utils/src/channels/Message';

import { NODE_TYPES } from '../constants/nodeTypes';
import { Config } from '../../config';
import {
  IS_MOUNTED_REQUEST,
  IS_MOUNTED_RESPONSE,
} from './messages';

const FIELD_STATUSES = {
  INIT: 'init',
  DESTROYED: 'destroyed',
};

/**
 * Base class for Field
 * @class
 * @extends {ElementsInstance}
 */
class Field {
  /**
   * @param {boolean} condition
   * @param {string} message
   */
  static invariant(condition, message) {
    invariant(condition, `[${this.name}] ${message}`);
  }

  static get STATUSES() {
    return FIELD_STATUSES;
  }

  /**
   * @param {HTMLElement} options.node
   * @param {string} options.channelId
   */
  constructor(options) {
    const node = options?.node;
    this.constructor.invariant(
      typeof node === 'object' && node.nodeType === NODE_TYPES.ELEMENT_NODE,
      'Node for field should be a HTMLElement',
    );

    this.status = FIELD_STATUSES.INIT;
    this.node = node;
    this.channelId = options?.channelId;
  }

  /**
   * Mount field
   * @param {HTMLElement} node
   */
  mount(node) {
    this.appendTo(node);
    this.openChannel();
    this.registerHandlers();
  }

  /**
   * Unmount field
   */
  unmount() {
    this.destroy();
  }

  /**
   * Append field node to parent node
   * @param {HTMLElement} parent - parent node
   */
  appendTo(parent) {
    if (this.status === FIELD_STATUSES.DESTROYED) {
      this.constructor.invariant(
        false,
        'Field is destroyed and can not be append to a parent node.',
      );
      return;
    }

    this.constructor.invariant(
      typeof parent === 'object' && parent.nodeType === NODE_TYPES.ELEMENT_NODE,
      'Parent node should be a HTMLElement node',
    );

    this.parent = parent;
    parent.appendChild(this.node);
  }

  /**
   * Open channel with master frame
   * @private
   */
  openChannel() {
    this.constructor.invariant(
      Boolean(this.channelId),
      'Field does not have specified "channelId" for opening channel',
    );

    this.fieldSlaveChannel = new SlaveChannel({
      channelId: this.channelId,
      targetOrigin: Config.sdkOrigin,
    });
    this.fieldSlaveChannel.connect();
  }

  /**
   * Register allowed messages by master frame and give response by this messages
   * @private
   */
  registerHandlers() {
    this.fieldSlaveChannel.subscribe(IS_MOUNTED_REQUEST, () => {
      this.fieldSlaveChannel.postMessage(
        new Message(IS_MOUNTED_RESPONSE, { success: true }),
      );
    });
  }

  /**
   * Destroy field and remove its from parent, if it is specified
   */
  destroy() {
    if (this.status === FIELD_STATUSES.DESTROYED) {
      this.constructor.invariant(
        false,
        'Field is destroyed and can not be destroyed again.',
      );
      return;
    }

    if (this.parent) {
      this.parent.removeChild(this.node);
    }
    this.status = FIELD_STATUSES.DESTROYED;
    this.node = null;
    this.parent = null;

    if (this.fieldSlaveChannel) {
      this.fieldSlaveChannel.destroy();
      this.fieldSlaveChannel = null;
    }
  }
}

export default Field;
export {
  Field,
};
