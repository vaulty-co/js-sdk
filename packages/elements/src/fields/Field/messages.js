const IS_MOUNTED_REQUEST = 'is_mounted_request';
const IS_MOUNTED_RESPONSE = 'is_mounted_response';

export {
  IS_MOUNTED_REQUEST,
  IS_MOUNTED_RESPONSE,
};
