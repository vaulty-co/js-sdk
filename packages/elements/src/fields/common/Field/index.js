import invariant from 'invariant';
import { BroadcastChannel } from 'broadcast-channel';
import { SlaveChannel } from '@js-sdk/utils/src/channels/SlaveChannel';
import { Message } from '@js-sdk/utils/src/channels/Message';
import { isSafari } from '@js-sdk/utils/src/helpers/isSafari';
import { Node } from '@js-sdk/utils/src/nodes/Node';
import { ComposedValidator } from '@js-sdk/utils/src/validators/ComposedValidator';

import { Config } from '../../../config';
import { VALIDATORS_TYPES } from '../../../validators/constants';
import { VALIDATORS_REGISTRY } from '../../../validators/registry';
import {
  FIELD_LOADED,
  INITIALIZE_FIELD_REQUEST,
  INITIALIZE_FIELD_RESPONSE,
  FIELD_DATA_CHANGE_RESPONSE,
  FIELD_FOCUS_CHANGE,
  FOCUS_FIELD,
  BLUR_FIELD,
  CLEAR_FIELD,
} from './messages';

const FIELD_STATUSES = {
  INIT: 'init',
  DESTROYED: 'destroyed',
};

/**
 * Base class for Field
 * @class
 * @extends {ElementsInstance}
 */
class Field {
  /**
   * @param {boolean} condition
   * @param {string} message
   */
  static invariant(condition, message) {
    invariant(condition, `[${this.name}] ${message}`);
  }

  static get STATUSES() {
    return FIELD_STATUSES;
  }

  /**
   * @param {FieldOptions} options
   */
  constructor(options) {
    const node = options?.node;
    this.constructor.invariant(
      typeof node === 'object' && node instanceof Node,
      'Field\'s node should be instance of Node',
    );

    this.status = FIELD_STATUSES.INIT;
    this.fieldNode = node;
    this.channelId = options?.channelId;
    this.composedValidator = new ComposedValidator();

    let broadcastChannelOptions = {};
    // Notice: Safari does not allow IndexDB fallback in iFrame and we should manually
    // use localStorage
    if (isSafari()) {
      broadcastChannelOptions = {
        type: 'localstorage',
      };
    }
    this.broadcastChannel = new BroadcastChannel(
      options?.sdkId,
      broadcastChannelOptions,
    );
  }

  /**
   * Mount field
   * @param {HTMLElement} node
   */
  mount(node) {
    this.appendTo(node);
    this.openChannel();
    this.registerHandlers();
  }

  /**
   * Unmount field
   */
  unmount() {
    this.destroy();
  }

  /**
   * Append field node to parent node
   * @param {HTMLElement} parent - parent node
   */
  appendTo(parent) {
    if (this.status === FIELD_STATUSES.DESTROYED) {
      this.constructor.invariant(
        false,
        'Field is destroyed and can not be append to a parent node.',
      );
      return;
    }

    this.fieldNode.appendTo(parent);
  }

  /**
   * Focus field
   */
  focusField() {
    this.fieldNode.focus();
  }

  /**
   * Blur field
   */
  blurField() {
    this.fieldNode.blur();
  }

  /**
   * Clear field
   */
  clearField() {
    this.fieldNode.clear();
  }

  /**
   * Destroy field and remove its from parent, if it is specified
   */
  destroy() {
    if (this.status === FIELD_STATUSES.DESTROYED) {
      this.constructor.invariant(
        false,
        'Field is destroyed and can not be destroyed again.',
      );
      return;
    }

    this.status = FIELD_STATUSES.DESTROYED;

    this.fieldNode.destroy();
    this.fieldNode = null;

    if (this.fieldSlaveChannel) {
      this.fieldSlaveChannel.destroy();
      this.fieldSlaveChannel = null;
    }

    this.broadcastChannel.close();
    this.broadcastChannel = null;
  }

  /**
   * Open channel with master frame
   * @private
   */
  openChannel() {
    this.constructor.invariant(
      Boolean(this.channelId),
      'Field does not have specified "channelId" for opening channel',
    );

    this.fieldSlaveChannel = new SlaveChannel({
      channelId: this.channelId,
      targetOrigin: Config.sdkOrigin,
    });
    this.fieldSlaveChannel.connect();
    this.fieldSlaveChannel.postMessage(
      new Message(FIELD_LOADED, {
        success: true,
      }),
    );
  }

  /**
   * Register allowed messages by master frame and give response by this messages
   * @private
   */
  registerHandlers() {
    this.fieldSlaveChannel.subscribe(INITIALIZE_FIELD_REQUEST, (message) => {
      const {
        payload: {
          id, name, style = {}, validators = [],
        },
      } = message;
      this.id = id;
      this.name = name;
      this.composedValidator = this.createValidators(validators);
      this.fieldNode.setStyle(style);
      this.fieldSlaveChannel.postMessage(
        new Message(INITIALIZE_FIELD_RESPONSE, { success: true }),
      );
    });
    this.fieldSlaveChannel.subscribe(FOCUS_FIELD, () => {
      this.focusField();
    });
    this.fieldSlaveChannel.subscribe(BLUR_FIELD, () => {
      this.blurField();
    });
    this.fieldSlaveChannel.subscribe(CLEAR_FIELD, () => {
      this.clearField();
    });
  }

  /**
   * Send field data changes
   * @param {FieldDataChangePayload} payload
   * @protected
   */
  sendDataChanges(payload) {
    this.fieldSlaveChannel.postMessage(
      new Message(
        FIELD_DATA_CHANGE_RESPONSE,
        payload,
      ),
    );
  }

  /**
   * Send field focus
   * @param {FieldFocusChangePayload} payload
   * @protected
   */
  sendFocusChanges(payload) {
    this.fieldSlaveChannel.postMessage(
      new Message(
        FIELD_FOCUS_CHANGE,
        payload,
      ),
    );
  }

  /**
   * Create validators by provided from library user validators types
   * @param {Array<string>} providedValidators
   * @returns {ComposedValidator}
   * @private
   */
  createValidators(providedValidators = []) {
    const validatorsTypesSetup = Object
      .keys(VALIDATORS_TYPES)
      .map((key) => VALIDATORS_TYPES[key])
      .filter((validatorType) => providedValidators.includes(validatorType));
    const validators = validatorsTypesSetup.map((validatorType) => {
      const Validator = VALIDATORS_REGISTRY[validatorType];
      this.constructor.invariant(
        Boolean(Validator),
        `Does not found out validator: '${validatorType}'`,
      );
      return new Validator();
    });

    return new ComposedValidator(validators);
  }
}

export default Field;
export {
  Field,
};
