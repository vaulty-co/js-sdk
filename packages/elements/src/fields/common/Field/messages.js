const INITIALIZE_FIELD_REQUEST = 'initialize_request';
const INITIALIZE_FIELD_RESPONSE = 'initialize_response';
const GET_FIELD_DATA_REQUEST = 'get_field_data_request';
const GET_FIELD_DATA_RESPONSE = 'get_field_data_response';
const FIELD_DATA_CHANGE_RESPONSE = 'FIELD_DATA_CHANGE_RESPONSE';

export {
  INITIALIZE_FIELD_REQUEST,
  INITIALIZE_FIELD_RESPONSE,
  GET_FIELD_DATA_REQUEST,
  GET_FIELD_DATA_RESPONSE,
  FIELD_DATA_CHANGE_RESPONSE,
};
