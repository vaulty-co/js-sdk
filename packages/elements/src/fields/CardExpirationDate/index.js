import cardValidator from 'card-validator';

import { CardExpirationDateNode } from './nodes/CardExpirationDateNode';
import { InputField } from '../common/InputField';

/**
 * It renders card expiration date input
 * @class
 */
class CardExpirationDate extends InputField {
  constructor(options) {
    const node = new CardExpirationDateNode();

    super({
      ...options,
      node,
    });
  }

  /**
   * Get data for sending its to controller
   * @returns {{id: string, name: FieldName, data: { month: string, year: string }}}
   */
  getData() {
    const { month, year } = cardValidator.expirationDate(
      this.fieldNode.getValue(),
    );
    return {
      id: this.id,
      name: this.name,
      data: {
        month,
        year,
      },
    };
  }
}

export default CardExpirationDate;
export {
  CardExpirationDate,
};
