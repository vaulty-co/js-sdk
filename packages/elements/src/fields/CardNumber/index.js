import { CardNumberNode } from './nodes/CardNumberNode';
import { InputField } from '../common/InputField';

/**
 * It renders simple text input node
 * @class CardNumberElement
 * @extends InputFieldElement
 */
class CardNumber extends InputField {
  constructor(options) {
    const node = new CardNumberNode();

    super({
      ...options,
      node,
    });
  }
}

export default CardNumber;
export {
  CardNumber,
};
