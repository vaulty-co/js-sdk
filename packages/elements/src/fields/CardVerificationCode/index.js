import { CardVerificationCodeNode } from './nodes/CardVerificationCodeNode';
import { InputField } from '../common/InputField';

/**
 * It renders card verification code input
 * @class
 */
class CardVerificationCode extends InputField {
  constructor(options) {
    const node = new CardVerificationCodeNode();

    super({
      ...options,
      node,
    });
  }
}

export default CardVerificationCode;
export {
  CardVerificationCode,
};
