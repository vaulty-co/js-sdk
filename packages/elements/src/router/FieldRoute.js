import { Route } from './Route';
import { Field } from '../fields/Field';

/**
 * Route for Field
 */
class FieldRoute extends Route {
  /**
   * Render route
   * @returns {Field}
   */
  render() {
    const field = this.renderFunction();
    this.constructor.invariant(
      (field instanceof Field),
      `Field route "${this.name}" does not return Field in render function.`,
    );
    return field;
  }
}

export default FieldRoute;
export {
  FieldRoute,
};
