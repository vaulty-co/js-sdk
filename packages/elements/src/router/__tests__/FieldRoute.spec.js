import { Field } from 'vaulty-js-sdk-elements/src/fields/Field';
import { FieldRoute } from '../FieldRoute';

describe('FieldRoute', () => {
  let route;
  let renderFunction;

  beforeEach(() => {
    renderFunction = jest.fn(() => {
      const input = document.createElement('input');
      input.type = 'text';
      return new Field(input);
    });
    route = new FieldRoute(
      'inputText',
      {
        field: 'input',
        type: 'text',
      },
      renderFunction,
    );
  });

  describe('#render', () => {
    it('should run render function', () => {
      route.render();

      expect(renderFunction).toBeCalled();
    });

    it('should throw exception, when it does not return Field instance', () => {
      route = new FieldRoute(
        'inputText',
        {
          field: 'input',
          type: 'text',
        },
        () => null,
      );

      expect(() => {
        route.render();
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
