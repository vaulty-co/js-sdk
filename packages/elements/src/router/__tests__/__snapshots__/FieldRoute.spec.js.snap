// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldRoute #render should throw exception, when it does not return Field instance 1`] = `"[FieldRoute] Field route \\"inputText\\" does not return Field in render function."`;
